radicle-httpd/src/api/v1/projects.rs

```diff
@@ -224,16 +227,15 @@ async fn activity_handler(
 /// `GET /projects/:project/tree/:sha/*path`
 async fn tree_handler(
     Extension(ctx): Extension<Context>,
-    Path((project, sha, path)): Path<(Id, Oid, String)>,
+    Path((_, sha, path)): Path<(Id, Oid, String)>,
 ) -> impl IntoResponse {
     let path = path.strip_prefix('/').ok_or(Error::NotFound)?.to_string();
     let storage = &ctx.profile.storage;
-    let repo = storage.repository(project)?;
-    let tree = radicle_surf::object::tree(&repo.raw().into(), Some(Sha { sha }), Some(path))?;
+    let repo = Repository::open(storage.path())?;
+    let sub_dir = repo.directory(sha, &path)?;
```

Comment:
I find the use of `directory` here really confusing, (and `file` in other places), since these are not directories on the file system, but trees inside the git object database..

Reply:
I think there's some cognitive dissonance there; eg. if it's a directory, I'd expect that I can interact with it like a directory, eg. move it or place files inside of it, or delete it.

Reply:
Good to know that! This is a read-only directory.  We can document that or name it better. However now I'm a bit concerned that maybe we should get rid of `Directory` and `File`, and only use `Tree` and `Blob`? 

For some reason, I automatically felt comfortable with the read-only nature of `Directory` and `File` provided by `radicle-surf` API. Didn't realize others would feel so differently.

Reply:
Well, it's still not quite a read-only directory, since it doesn't exist on the file system; ie. I can't `cd` into it, and I can't access it via `std::fs` or check its file-system path. It's some kind of "virtual" directory called a "tree", so I'd just name it what it is to avoid confusing it with a regular directory.. Especially that Git does also work with regular directories, in working copies, and surf may want to provide an API for that at some point.

Reply:
I think the reason I liked the name `directory` (and `file`) is that I felt they naturally fit with the motivation/ideas of `radicle-surf`:

"radicle-surf is a library to describe a Git repository as a file system. It aims to provide an easy-to-use API to browse a repository via the concept of files and directories for any given revision." ([README](https://github.com/radicle-dev/radicle-git/tree/main/radicle-surf)).

It is surely possible to only use `Tree` (and `Blob`) in `radicle-surf` and let surf's users to convert them into `Directory` and `File` when presenting to the end user. But I thought it's a good idea for `radicle-surf` to present `Directory` and `File` in its API already.

I'm a bit surprised to see you mention ` I can't cd into it, and I can't access it via std::fs or check its file-system path. `.  as that's not what I expected of `Directory` from such a library. I was wondering if other users think the same way. I will try to ask around more.

Regarding the possible future of supporting working copies, I think it would need some additional design and it's difficult for me to predict anything: will it support write? will it be part of surf? My take is that I don't think having `Directory` will prevent developing that feature.

Reply:
> "radicle-surf is a library to describe a Git repository as a file system. It aims to provide an easy-to-use API to browse a repository via the concept of files and directories for any given revision." ([README](https://github.com/radicle-dev/radicle-git/tree/main/radicle-surf)).

That was written a long time ago, when surf was multi-vcs, which means that concepts like Blob or Tree would not work, since not all VCSs have those.

> I'm a bit surprised to see you mention I can't cd into it, and I can't access it via std::fs or check its file-system path.. as that's not what I expected of Directory from such a library. I was wondering if other users think the same way. I will try to ask around more.

To flip it around: I *wouldn't* expect that an object called "Directory" would have a SHA-1; yet in this case it does. However you put it, the expectations are broken due to the mismatch between Directories and Trees. When working with bare repositories, I would search the docs for "blob" and "tree", given that's the object I want to work with and be very surprised if I don't get any results in a git-oriented library. If I saw "directory" in the docs, I would expect that to be used for actual file-system objects, eg. repository directories, user config directories or the working copy directories.

Reply:
I find it really strange that we would shy away from naming something what it actually is when there is a specific and well-known term used for that thing. That alone I think would create confusion: "why did they call it File? Is it different in some way from a Blob? Does it only work for blobs in working copies?" The only good reason to pick a different name in this case is if it is *not* intended to behave like a Blob/Tree.

Reply:
I think it's a good discussion that helps me think more about the names. (As they say, naming is hard ;-)) . I spent a bit time to read the git [book](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects). My take is that, in git world:

1. `Tree` and `Blob` are Git internals, even though they are very familiar terms for developers of `radicle-git` / `hearwood`.

2. `Tree` and `Blob` are contents.  `Tree` is the content of a virtual directory, and `Blob` is the content of a file. This means that if two directories have the same content, they share the same tree even when they are different directories.  Likewise, two files can share the same blob.

For example in a simple test: 
```
$ git cat-file -p branch_b^{tree}
040000 tree 494c44e7fe186a29d2f360a6d5a4071a47a8041c    dir_1
040000 tree 494c44e7fe186a29d2f360a6d5a4071a47a8041c    dir_2
100644 blob a5bce3fd2565d8f458555a0c6f42d0504a848bd5    test1
100644 blob f83bb2683ace630ef6e8dcfde2aedde6948c90b4    test2
100644 blob f83bb2683ace630ef6e8dcfde2aedde6948c90b4    test3
```

This seems to me, logically a `Directory` is a container that has a `Tree` (its content) and other info (mainly its path). `File` is a container that has a `Blob` (its content) and other info (file name). 

And that's how I understand the difference between (`Directory`, `File`) and (`Tree`, `Blob`). 

3.  If we expose `Tree` and `Blob` in radicle-surf's public API, I hope they map to what they mean in git. Currently they are not, for example  [`Blob`](https://github.com/radicle-dev/radicle-git/blob/main/radicle-surf/src/source/object/blob.rs#L36) contains a `File`. I will play around this area in radicle-surf to see if I can do anything.

Reply:
> This seems to me, logically a Directory is a container that has a Tree (its content) and other info (mainly its path). File is a container that has a Blob (its content) and other info (file name).

This is true in a sense but also confuses things: git doesn't have directories or files in its object database, and the file system doesn't have trees/blobs. So saying that a directory "contains" a tree is not right, unless you mean it in a very abstract sense. I'd say more that a tree is *associated* with a directory.

Reply:
> I'd say more that a tree is associated with a directory.

Or a tree can materialize a directory, i.e. when you checkout a commit the associated tree is rendered into a file system directory.

Reply:
Another note that I'll repeat here is that the httpd currently requires additional information for these types:
* What was the last commit for a Blob or Tree
* The kind of the Blob content, i.e. plain text or binary
* The entries for the current Tree

These are all computable so we either provide utility functions for calculating those or we provide the data itself.

Reply:
I have updated https://github.com/radicle-dev/radicle-git/pull/79 to try out using `Tree` (and later will finish `Blob` part). Will update here after that.

Reply:
I have updated to use `Tree` and `Blob` here.  They are content based, same as in git.  `Tree` itself does not contain a path, and `Blob` itself does not contain a name.  For `path` and `name`, the caller should already have them.